### PR TITLE
 Make bucket with builder

### PR DESCRIPTION
**Pull request by sinhaashish**

### CODE REVIEW

File: api/src/main/java/io/minio/MinioClient.java [hunk 1]
```diff
@@ -3324,13 +3326,55 @@ public void makeBucket(String bucketName, String region)
...
+        new MakeBucketArgs()
+            .builder()
+            .bucket(bucketName)
+            .region(region)
+            .objectLock(objectLock)
+            .build());
+  }
+
+  /**
+   * Creates a bucket with make bucket arguments.
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java:3345]: `Creates a bucket with region and object lock.`

File: api/src/main/java/io/minio/MinioClient.java [hunk 2]
```diff
@@ -3324,13 +3326,55 @@ public void makeBucket(String bucketName, String region)
...
+   *
+   * <pre>Example:{@code
+   * minioClient.makeBucket(MakeBucketArgs args);
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java:3348]: You would need to provide examples as similar as `API.md` here.

File: docs/API.md [hunk 3]
```diff
@@ -645,55 +645,30 @@ for (Result<Item> result : results) {
...
-| ``objectLock`` | _boolean_ | Flag to enable object lock feature.                    |
+| Parameter      | Type             | Description                |
+|:---------------|:-----------------|:---------------------------|
+| ``args``       | _[MakeBucketArgs]_ | Arguments to create bucket |
 
 __Example__
 ```java
-minioClient.makeBucket("my-bucketname", "us-west-2", true);
+// Create bucket with default region.
+minioClient.makeBucket(new MakeBucketArgs.Builder().bucket("my-bucketname").build());
```
Comment (balamurugana) [inline docs/API.md:661]: Its better to have new line after each example.

File: docs/API.md [hunk 4]
```diff
@@ -645,55 +645,30 @@ for (Result<Item> result : results) {
...
+// Create bucket with specific region.
+minioClient.makeBucket(
+    new MakeBucketArgs.Builder().bucket("my-bucketname").region("us-west-1").build());
```
Comment (balamurugana) [inline docs/API.md:664]: Make same indention like below example.

File: docs/API.md [hunk 5]
```diff
@@ -645,55 +645,30 @@ for (Result<Item> result : results) {
...
-
-__Parameters__
-| Parameter      | Type      | Description                                            |
-|:---------------|:----------|:-------------------------------------------------------|
-| ``bucketName`` | _String_  | Name of the bucket.                                    |
-| ``region``     | _String_  | (Optional) Region in which the bucket will be created. |
-| ``objectLock`` | _boolean_ | Flag to enable object lock feature.                    |
+| Parameter      | Type             | Description                |
+|:---------------|:-----------------|:---------------------------|
+| ``args``       | _[MakeBucketArgs]_ | Arguments to create bucket |
```
Comment (balamurugana) [inline docs/API.md:656]: You would need to add link to `[MakeBucketArgs]`

File: examples/MakeBucketWithLock.java [hunk 6]
```diff
@@ -36,7 +37,8 @@ public static void main(String[] args)
         System.out.println("my-bucketname already exists");
       } else {
         // Create bucket 'my-bucketname' with object lock functionality enabled
-        s3Client.makeBucket("my-bucketname", null, true);
+        s3Client.makeBucket(
+            new MakeBucketArgs.Builder().bucket("my-bucketname").objectLock(true).build());
```
Comment (balamurugana) [inline examples/MakeBucketWithLock.java:41]: Now you could have one example `MakeBucket.java` and add three variation there.

File: functional/FunctionalTest.java [hunk 7]
```diff
@@ -251,22 +251,21 @@ public static void makeBucket_test1() throws Exception {
     }
   }
 
-  /** Test: makeBucket(String bucketName, String region). */
+  /** Test: makeBucket(MakeBucketArgs args). */
   public static void makeBucketwithRegion_test() throws Exception {
     if (!mintEnv) {
-      System.out.println("Test: makeBucket(String bucketName, String region)");
+      System.out.println("Test: makeBucket(MakeBucketArgs args)");
```
Comment (balamurugana) [inline functional/FunctionalTest.java:257]: You can add a tag like `with region`, `with objectLock` or `with region and objectLock`

File: api/src/main/java/io/minio/MakeBucketArgs.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+  public boolean objectLock() {
+    return objectLock;
+  }
+
+  public MakeBucketArgs() {
+    this(new Builder());
+  }
+
+  public Builder builder() {
```
[the remark this answers is not in the document]
Comment (sinhaashish) [reply api/src/main/java/io/minio/MakeBucketArgs.java:35]: You can find @anjalshireesh  comment above, and thus i have made this as static and created the object like 
```
MakeBucketArgs.builder()
        .bucket("<bucketname>")
        .region("<region>")
        .objectLock(<true/false>)
        .build()
```

File: docs/API.md (hunk 5 above)
Comment (sinhaashish) [reply docs/API.md:656]: i have tried the below but the link is visible. Is this correct approach
```

__Parameters__

| Parameter      | Type               | Description                |
|:---------------|:-------------------|:---------------------------|
| ``args``       | _[MakeBucketArgs](#https://minio.github.io/minio-java/io/minio/MakeBucketArgs.html)_ | Arguments to create bucket |

```

File: api/src/main/java/io/minio/BucketArgs.java [hunk 9]
```diff
@@ -16,30 +16,55 @@
 
 package io.minio;
 
+/** Bucket Arguments to hold base bucket properties */
```
Comment (balamurugana) [inline api/src/main/java/io/minio/BucketArgs.java:19]: `/** Base argument class holds bucket name and region */`

File: api/src/main/java/io/minio/BucketArgs.java [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+public abstract class BucketArgs {
+  private final String name;
+  private final String region;
+
+  BucketArgs(Builder<?> builder) {
+    this.name = builder.name;
+    this.region = builder.region;
+  }
+
+  /** Returns the name of bucket */
```
Comment (balamurugana) [inline api/src/main/java/io/minio/BucketArgs.java:29]: `/** Returns bucket name */` to consistent with other docs.

File: api/src/main/java/io/minio/BucketArgs.java [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+  public String bucketName() {
+    return name;
+  }
+
+  /** Returns the region of bucket */
```
Comment (balamurugana) [inline api/src/main/java/io/minio/BucketArgs.java:34]: `/** Returns region */`

File: api/src/main/java/io/minio/BucketArgs.java [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+    public Builder() {}
+
+    public Builder(BucketArgs args) {
+      this.name = args.bucketName();
+      this.region = args.region();
+    }
+
+    @SuppressWarnings("unchecked")
+    /** Its safe to type cast to T as T is inherited this class. */
```
Comment (balamurugana) [inline api/src/main/java/io/minio/BucketArgs.java:52]: Please check how to add justification in `@suppressWarnings`

File: api/src/main/java/io/minio/BucketArgs.java [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+    public T region(String region) {
+      this.region = region;
+      return (T) this;
+    }
+
+    /** Validate the name of the bucket */
+    public void validateName(String name) {
```
Comment (balamurugana) [inline api/src/main/java/io/minio/BucketArgs.java:67]: We could make this as `static` method.

File: api/src/main/java/io/minio/MakeBucketArgs.java [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.minio;
+
+/** Bucket Arguments to be used to create Bukcet. */
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MakeBucketArgs.java:19]: `/** Argument class of {@see makeBucket()}. */`

File: api/src/main/java/io/minio/MakeBucketArgs.java [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+public class MakeBucketArgs extends BucketArgs {
+  private final boolean objectLock;
+
+  private MakeBucketArgs(Builder builder) {
+    super(builder);
+    this.objectLock = builder.objectLock;
+  }
+
+  /** Return true if object lock functionality is enabled */
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MakeBucketArgs.java:28]: Just `/** Returns object lock flag. */

File: api/src/main/java/io/minio/BucketArgs.java [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+  public String bucketName() {
+    return name;
+  }
+
+  /** Returns the region of bucket */
+  public String region() {
+    return region;
+  }
+
+  /** Builder class to create base bucket object */
```
Comment (balamurugana) [inline api/src/main/java/io/minio/BucketArgs.java:39]: `/** Base argument builder class. */` is sufficient.

File: api/src/main/java/io/minio/MakeBucketArgs.java [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+  public boolean objectLock() {
+    return objectLock;
+  }
+
+  /** Return new Builder */
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public static final class Builder extends BucketArgs.Builder<Builder> {
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MakeBucketArgs.java:38]: /** Argument builder of {@see makeBucket()}. */`

File: api/src/main/java/io/minio/MakeBucketArgs.java [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+    private boolean objectLock;
+
+    public Builder() {}
+
+    public Builder(MakeBucketArgs args) {
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MakeBucketArgs.java:43]: You could remove `Builder(XXXArgs args)` constructor everywhere and will be added on demand.

File: docs/API.md [hunk 19]
```diff
@@ -645,55 +645,40 @@ for (Result<Item> result : results) {
...
-Creates a bucket with given region.
-
-__Parameters__
-| Parameter      | Type     | Description                                 |
-|:---------------|:---------|:--------------------------------------------|
-| ``bucketName`` | _String_ | Name of the bucket.                         |
-| ``region``     | _String_ | Region in which the bucket will be created. |
+| Parameter      | Type               | Description                |
+|:---------------|:-------------------|:---------------------------|
+| ``args``       | _[MakeBucketArgs](#https://minio.github.io/minio-java/io/minio/MakeBucketArgs.html)_ | Arguments to create bucket |
```
Comment (balamurugana) [inline docs/API.md:657]: Check how `[Method]` link is added as reference.

File: examples/MakeBucket.java [hunk 20]
```diff
@@ -37,14 +38,36 @@ public static void main(String[] args)
...
+      } else {
+
+        // Create bucket 'my-bucketname3' , egion and object lock functionality enabled.
+        minioClient.makeBucket(
+            MakeBucketArgs.builder()
+                .bucket("my-bucketname3")
+                .region("us-west-1")
+                .objectLock(true)
+                .build());
+        System.out.println("my-bucketname3 is created successfully");
```
Comment (balamurugana) [inline examples/MakeBucket.java:69]: 1. You could have bucket names as `my-bucketname`, `my-bucketname-in-eu`, `my-bucketname-in-eu-with-object-lock`
2. I prefer longer block in `if` i.e. `if (!minioClient.bucketExists(...)) {` is better.

File: functional/FunctionalTest.java [hunk 21]
```diff
@@ -251,22 +251,21 @@ public static void makeBucket_test1() throws Exception {
     }
   }
 
-  /** Test: makeBucket(String bucketName, String region). */
+  /** Test: makeBucket(MakeBucketArgs args). */
   public static void makeBucketwithRegion_test() throws Exception {
     if (!mintEnv) {
-      System.out.println("Test: makeBucket(String bucketName, String region)");
+      System.out.println("Test: makeBucket(MakeBucketArgs args) with region");
```
Comment (balamurugana) [inline functional/FunctionalTest.java:257]: Add tag before the API `Test: with region: makeBucket(...)`

File: functional/FunctionalTest.java [hunk 22]
```diff
@@ -303,6 +301,33 @@ public static void makeBucketWithPeriod_test() throws Exception {
     }
   }
 
+  /** Test: makeBucket(MakeBucketArgs args). */
+  public static void makeBucketwithRegioAndObjectLock() throws Exception {
```
Comment (balamurugana) [inline functional/FunctionalTest.java:305]: I think we should use `makeBucket_testN()` like other tests.

File: build.gradle [hunk 23]
```diff
@@ -67,6 +67,7 @@ subprojects {
         it.options.fork = true
         it.options.compilerArgs += ["-Xlint:unchecked", "-Xlint:deprecation", "-Xlint:-options"]
         it.options.encoding = "UTF-8"
+        it.options.compilerArgs << "-Werror"
```
Comment (balamurugana) [inline build.gradle:70]: Can we add `"-Werror"` in `it.options.compilerArgs +=` line?

Comment (sinhaashish) [reply build.gradle]: This worked
`it.options.compilerArgs += ["-Xlint:unchecked", "-Xlint:deprecation", "-Xlint:-options", "-Werror"]`

File: api/src/main/java/io/minio/MakeBucketArgs.java [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+      this.name = args.bucketName();
+      this.region = args.region();
```
[the remark this answers is not in the document]
Comment (sinhaashish) [reply api/src/main/java/io/minio/MakeBucketArgs.java]: This constructor has been removed as per @balamurugana  comment
> You could remove Builder(XXXArgs args) constructor everywhere and will be added on demand.

**Pull request merged**